crates/bevy_ecs/macros/src/fetch.rs

```diff
@@ -0,0 +1,622 @@
+    let read_only_fetch_struct_name = if has_mutable_attr {
+        Ident::new(&format!("{}ReadOnlyFetch", struct_name), Span::call_site())
+    } else {
+        fetch_struct_name.clone()
+    };
+
+    let fields = match &ast.data {
+        Data::Struct(DataStruct {
+            fields: Fields::Named(fields),
+            ..
+        }) => &fields.named,
+        _ => panic!("Expected a struct with named fields"),
+    };
+
+    let mut phantom_field_idents = Vec::new();
+    let mut phantom_field_types = Vec::new();
+    let mut field_idents = Vec::new();
+    let mut field_types = Vec::new();
+    let mut query_types = Vec::new();
+    let mut fetch_init_types = Vec::new();
+
+    for field in fields.iter() {
+        let (world_lifetime, fetch_lifetime) = match (&world_lifetime, &fetch_lifetime) {
+            (GenericParam::Lifetime(world), GenericParam::Lifetime(fetch)) => {
+                (&world.lifetime, &fetch.lifetime)
+            }
+            _ => unreachable!(),
+        };
+        let WorldQueryFieldTypeInfo {
+            query_type,
+            fetch_init_type: init_type,
+            is_phantom_data,
+        } = read_world_query_field_type_info(&field.ty, world_lifetime, fetch_lifetime);
+
+        let field_ident = field.ident.as_ref().unwrap().clone();
+        if is_phantom_data {
+            phantom_field_idents.push(field_ident.clone());
+            phantom_field_types.push(field.ty.clone());
+        } else {
+            field_idents.push(field_ident.clone());
+            field_types.push(field.ty.clone());
+            query_types.push(query_type);
+            fetch_init_types.push(init_type);
+        }
+    }
+
+    FetchImplTokens {
+        struct_name,
+        struct_name_read_only,
+        fetch_struct_name,
+        state_struct_name,
+        read_only_fetch_struct_name,
+        fetch_trait_punctuated_lifetimes,
+        impl_generics,
+        ty_generics,
+        where_clause,
+        has_mutable_attr,
+        has_world_lifetime,
+        world_lifetime,
+        state_lifetime,
+        fetch_lifetime,
+        phantom_field_idents,
+        phantom_field_types,
+        field_idents,
+        field_types,
+        query_types,
+        fetch_init_types,
+    }
+}
+
+struct WorldQueryFieldTypeInfo {
+    /// We convert `Mut<T>` to `&mut T` (because this is the type that implements `WorldQuery`)
+    /// and store it here.
+    query_type: Type,
+    /// The same as `query_type` but with `'fetch` lifetime.
+    fetch_init_type: Type,
+    is_phantom_data: bool,
+}
+
+fn read_world_query_field_type_info(
+    ty: &Type,
+    world_lifetime: &Lifetime,
+    fetch_lifetime: &Lifetime,
+) -> WorldQueryFieldTypeInfo {
+    let path = bevy_ecs_path();
+
+    let query_type = parse_quote!(<#ty as #path::query::FetchedItem>::Query);
+    let mut fetch_init_type = parse_quote!(<#ty as #path::query::FetchedItem>::Query);
```

Comment:
No idea of the result of `parse_quote` implements `Clone`, but if it does, then this seems clearer to me.
```suggestion
    let query_type = parse_quote!(<#ty as #path::query::FetchedItem>::Query);
    let mut fetch_init_type = query_type.clone();
```